bluefog/common/topology_util.py

```diff
@@ -303,6 +303,79 @@ def FullyConnectedGraph(size: int) -> nx.DiGraph:
     return G
 
 
+def InnerOuterRingGraph(world_size: int, local_size: int) -> nx.DiGraph:
+    """Generate Inner Ring and Outer Ring (Unilateral) Graph.
+
+    Within one machine all inner rank/processes is connected in ring and all
+    nodes with same local size across all machines is connected through another ring.
+
+    .. plot::
+
+        >>> import networkx as nx
+        >>> from bluefog.common import topology_util
+        >>> G = topology_util.InnerOuterRingGraph(12, 3)
+        >>> nx.draw_circular(G)
+    """
+    total_nodes = world_size
+    num_machines = world_size // local_size
+    nodes_per_machine = local_size
+    assert world_size % local_size == 0, "It should be used under homogeneous environment only."
+
+    topo = np.zeros([total_nodes, total_nodes])
+
+    for i in range(total_nodes):
+        for j in range(total_nodes):
+            machine_i, local_rank_i = i // nodes_per_machine, i % nodes_per_machine
+            machine_j, local_rank_j = j // nodes_per_machine, j % nodes_per_machine
+            if machine_i == machine_j:
+                topo[i, j] = 1
```

Comment:
Right, it is fuly connected. It was because of the subgraph requirement in the dynamic topology. Will remove this entire function after removing that requirement. (Temporarily add the comments for that)